sdk/storage/azure-storage-blob-batch/src/main/java/com/azure/storage/blob/batch/BlobBatchHelper.java

```diff
@@ -0,0 +1,137 @@
+    static Mono<SimpleResponse<Void>> mapBatchResponse(BlobBatch batch, ServicesSubmitBatchResponse batchResponse,
+        boolean throwOnAnyFailure, ClientLogger logger) {
+        /*
+         * Content-Type will contain the boundary for each batch response. The expected format is:
+         * "Content-Type: multipart/mixed; boundary=batchresponse_66925647-d0cb-4109-b6d3-28efe3e1e5ed"
+         */
+        String contentType = batchResponse.getDeserializedHeaders().getContentType();
+
+        // Split on the boundary [ "multipart/mixed; boundary", "batchresponse_66925647-d0cb-4109-b6d3-28efe3e1e5ed"]
+        String boundary = contentType.split("=", 2)[1];
+
+        return FluxUtil.collectBytesInByteBufferStream(batchResponse.getValue())
+            .flatMap(byteArrayBody -> {
+                String body = new String(byteArrayBody, StandardCharsets.UTF_8);
+
+                // Split the batch response body into batch operation responses.
+                for (String subResponse : body.split("--" + boundary)) {
+                    // This is a split value that isn't a response.
+                    if (!subResponse.contains("application/http")) {
+                        continue;
+                    }
+
+                    // The batch operation response will be delimited by two new lines.
+                    String[] subResponseSections = subResponse.split(HTTP_NEWLINE + HTTP_NEWLINE);
+
+                    // The first section will contain batching metadata.
+                    BlobBatchOperationResponse<?> batchOperationResponse =
+                        getBatchOperation(batch, subResponseSections[0], logger);
+
+                    // The second section will contain status code and header information.
+                    setStatusCodeAndHeaders(batchOperationResponse, subResponseSections[1]);
+
+                    // The third section will contain the body.
+                    if (subResponseSections.length > 2) {
+                        // The body is optional and may not exist.
+                        setBodyOrPotentiallyThrow(batchOperationResponse, subResponseSections[2], throwOnAnyFailure,
+                            logger);
+                    }
+                }
+
+                return Mono.just(new SimpleResponse<>(batchResponse, null));
+            });
+    }
+
+    private static BlobBatchOperationResponse<?> getBatchOperation(BlobBatch batch, String responseBatchInfo,
+        ClientLogger logger) {
+        Matcher contentIdMatcher = CONTENT_ID_PATTERN.matcher(responseBatchInfo);
+
+        int contentId;
+        if (contentIdMatcher.find()) {
+            contentId = Integer.parseInt(contentIdMatcher.group(1));
+        } else {
+            throw logger.logExceptionAsError(
+                new IllegalStateException("Batch operation response doesn't contain a 'Content-Id' header."));
+        }
+
+        return batch.getBatchRequest(contentId).setResponseReceived();
```

Comment:
>.setResponseReceived() [](start = 47, length = 22)

This should be called after setBodyOrPotentiallyThrow to ensure that all the fields are filled in before it is marked as valid for reading.

Reply:
Given this is more of an implementation detail I will clean this up in Preview 5